framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,311 @@
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.QualifierHierarchy;
+import org.checkerframework.framework.type.treeannotator.ListTreeAnnotator;
+import org.checkerframework.framework.type.treeannotator.TreeAnnotator;
+import org.checkerframework.framework.util.MultiGraphQualifierHierarchy;
+import org.checkerframework.framework.util.MultiGraphQualifierHierarchy.MultiGraphFactory;
+import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.TreeUtils;
+import org.checkerframework.javacutil.UserError;
+
+/**
+ * An annotated type factory for an accumulation checker.
+ *
+ * <p>New accumulation checkers should extend this class and implement a constructor, which should
+ * take a {@link BaseTypeChecker} and call both the constructor defined in this class and {@link
+ * #postInit()}.
+ */
+public abstract class AccumulationAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    /**
+     * The canonical top annotation for this accumulation checker: an instance of the accumulator
+     * annotation with no arguments.
+     */
+    public final AnnotationMirror top;
+
+    /** The canonical bottom annotation for this accumulation checker. */
+    public final AnnotationMirror bottom;
+
+    /**
+     * The annotation that accumulates things in this accumulation checker. Must be an annotation
+     * with exactly one field named "value" whose type is a String array.
+     */
+    private final Class<? extends Annotation> accumulator;
+
+    /**
+     * Create an annotated type factory for an accumulation checker.
+     *
+     * @param checker the checker
+     * @param accumulator the accumulator type in the hierarchy. Must be an annotation with a single
+     *     argument named "value" whose type is a String array.
+     * @param bottom the bottom type in the hierarchy, which must be a subtype of {@code
+     *     accumulator}. The bottom type should be an annotation with no arguments.
+     */
+    protected AccumulationAnnotatedTypeFactory(
+            BaseTypeChecker checker,
+            Class<? extends Annotation> accumulator,
+            Class<? extends Annotation> bottom) {
+        super(checker);
+
+        this.accumulator = accumulator;
+
+        // Check that the requirements of the accumulator are met.
+        Method[] accDeclaredMethods = accumulator.getDeclaredMethods();
+        if (accDeclaredMethods.length != 1) {
+            rejectMalformedAccumulator("not have more than one argument");
+        }
+        Method value = accDeclaredMethods[0];
+        if (value.getName() != "value") {
+            rejectMalformedAccumulator("name its argument \"value\"");
+        }
+        if (!value.getReturnType().isInstance(new String[0])) {
+            rejectMalformedAccumulator("have an argument of type String[]");
```

Comment:
This check, and the next one, are not about the annotation but about its `value` element, so the string created by `rejectMalformedAccumulator` is imprecise and possibly confusing.